### DISCUSSION TITLE
Django | ORM

### DISCUSSION THREAD

**Discussion by willkotheimer** [Ascend-Apollo-Apprentice-2024/learning-platform-api#24] [General]
6/12/24 - 

Api call to get all the students for the cohort is currently taking too long to respond so we began the investigation of reducing the response time.
1. We began on student view on line 138, and ended on 188 after the serializer assignment.
2. We learned about Django's ORM and how to retrieve records from the ORM using `Class.objects.method` 
3. We learned about how Django creates foreign key relationships in related properties using dunderscore , i.e. `students.filter(assigned_cohorts__cohort=cohort_filter)`

**Comment by jonathanvanduyne** [comment]
6/13/24 - 

Lazy initialization is running tons of instructions to the ORM prior to the query actually executing to fetch the data that we are requesting. Which means we are running a lot of queries prior to serialization with all of the connected tables. Running all of the queries separately is an expensive process that we need to solve. The best approach for now is to work from the client JSON structure to what it is receiving and work backwards to see how we can consolidate all of these queries into a few, hopefully one query.

**Comment by Odigene12** [comment]
6/13/24 -

To further clarify the relationship between classes, we know that we have the `NssUser` class connects to the `Cohort` class through a join table called `NssUserCohort`, which is represented as a class that has a foreign key relationship with `NssUser` and `Cohort`. 

In the NssUserCohort class, the line `nss_user = models.ForeignKey("NssUser", on_delete=models.DO_NOTHING, related_name="assigned_cohorts")` creates the foreign key relationship and the `related_name` attribute is what gets used to then access the `Cohort` class from the `NssUser` class in the line `students.filter(assigned_cohorts__cohort=cohort_filter)`.

**Comment by jonathanvanduyne** [comment]
6/19/24 - 
Dug into the JSON received on the client and traced it back to the join tables between our data. Ended on MicroStudents serializer, moving forward we'll look at optimizing the join tables to see if we are duplicating work to get the same information that is adding cost resources to our query. Begin with looking at the student_view where we reinitialize students in our list creation, begin with the micro_students again

**Comment by Odigene12** [comment]
Now that we have created a SQL Function we know the information that it is returning (along with its data types for each column). We need to create a custom serializer to provide a JSON response to a client.
